### PR TITLE
librbd/cache/pwl: Using BlockGuard control overlap ops order when flu…

### PR DESCRIPTION
**Pull request by majianpeng**
…sh to osd.

In process of tests, we met some inconsistent-data problem. Test case
mainly use write,then discard to detect data consistent.
W/o pwl, write/discard are synchronous ops. After write, data already
located into osd. But w/ pwl, we use asynchronous api to send ops to
osd.
Although we mare sure send order. But send-order don't makre sure
complete order. This mean pwl keep order of write/discard. But it
don't keep the same semantics which use synchronous api. W/ pwl, it make
synchronous to asynchronous. For normal ops, it's not problem. But if
connected-commands w/ overlap, it make data inconsistent.
So we use BlockGuard to solve this issue.

Fixes: https://tracker.ceph.com/issues/49876
Fixes: https://tracker.ceph.com/issues/53108

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
@ideepika  @idryomov  run >3 hours for ssd/rwl cache, https://tracker.ceph.com/issues/49876/ https://tracker.ceph.com/issues/53108 both don't meet problems.
please review. Thanks!

**Comment by ideepika** [comment]
cc @orozery

**Comment by ideepika** [comment]
@majianpeng as per your mail, does using blockguard impact pwl performance significantly?

**Comment by majianpeng** [comment]
@ideepika  @ir

> @majianpeng as per your mail, does using blockguard impact pwl performance significantly?

No time for performance testing yet. Performance test today

**Comment by ir** [comment]
> @ideepika @ir
> 
> > @majianpeng as per your mail, does using blockguard impact pwl performance significantly?
> 
> No time for performance testing yet. Performance test today

:heart_eyes:

**Comment by ideepika** [comment]
@majianpeng one idea that Josh and I were discussing was to maintain a buffer of io's along with their metadata to maintain order, have you already explored that idea?

**Comment by majianpeng** [comment]
> @majianpeng one idea that Josh and I were discussing was to maintain a buffer of io's along with their metadata to maintain order, have you already explored that idea?

@ideepika   My currently implementation is use BlockGuard to fix overlap ops problems.  So can you explain in detain for Josh & you's method? THanks!

**Comment by ideepika** [comment]
> > @majianpeng one idea that Josh and I were discussing was to maintain a buffer of io's along with their metadata to maintain order, have you already explored that idea?
> 
> @ideepika My currently implementation is use BlockGuard to fix overlap ops problems. So can you explain in detain for Josh & you's method? THanks!

Discussed with Mykola a bit, if this does not hamper performance, the approach seems to fine solving these bugs for now.

**Comment by majianpeng** [comment]
@ideepika  . i revert patch 76f4d29d92be3f9. This will decrease performance.
W/ two patch,  i tested one image size=100G, cachesize=8G, 
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jianpeng/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jianpeng/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



QD/IOPS(k) | 1 | 8 | 16 | 32
-- | -- | -- | -- | --
patch+revert | 7.9 | 18.6 | 19.1 | 17.9
re-master | 8.3 | 19.8 | 19.5 | 18.9
patch/master | 95% | 94% | 98% | 95%



</body>

</html>
Looks performance don't drop much.

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
retest this please

File: src/librbd/cache/pwl/AbstractWriteLog.cc [hunk 1]
```diff
@@ -1672,7 +1688,24 @@ Context* AbstractWriteLog<I>::construct_flush_entry(std::shared_ptr<GenericLogEn
...
+	BlockGuardCell *detained_cell = nullptr;
+
+	std::lock_guard locker{m_flush_guard_lock};
+	m_flush_guard.release(log_entry->m_cell, &block_reqs);
+
+	for (auto &req : block_reqs) {
+	  m_flush_guard.detain(req.block_extent, &req, &detained_cell);
+	  if (detained_cell) {
+	    req.guard_ctx->cell = detained_cell;
+	    m_work_queue.queue(req.guard_ctx);
```
Comment (majianpeng) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:1704]: it need use m_image_ctx.op_work_queue

File: src/librbd/cache/pwl/AbstractWriteLog.cc [hunk 2]
```diff
@@ -1627,21 +1630,34 @@ bool AbstractWriteLog<I>::can_flush_entry(std::shared_ptr<GenericLogEntry> log_e
...
+  auto req = GuardedRequest(extent, guarded_ctx, false);
+  BlockGuardCell *cell = nullptr;
+
+  {
+    std::lock_guard locker(m_flush_guard_lock);
+    m_flush_guard.detain(req.block_extent, &req, &cell);
+  }
+  if (cell) {
+    req.guard_ctx->cell = cell;
+    m_work_queue.queue(req.guard_ctx);
```
Comment (majianpeng) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:1653]: this need m_image_ctx.op_work_queue.

**Comment by ideepika** [comment]
jenkins make check arm64

**Comment by ideepika** [comment]
jenkins retest api

**Comment by ideepika** [comment]
jenkins test make check arm64

**Comment by ideepika** [comment]
jenkins test api

**Review by ideepika — APPROVED** [review]
code lgtm @trociny would value your insights too, whenever feasible for you to take a look, thanks!

**Review by sunnyku — APPROVED** [review]
LGTM

**Review by trociny — APPROVED** [review]
LGTM, needs good testing though

**Comment by ideepika** [comment]
https://pulpito.ceph.com/ideepika-2021-12-02_12:23:25-rbd-wip-deepika-testing-2021-12-02-1334-distro-default-smithi/
https://tracker.ceph.com/issues/53112 [nfs mount issue]
[.MaxWhileTries: reached maximum tries (90) after waiting for 540 seconds]
https://tracker.ceph.com/issues/53138 [pg degrade]
https://tracker.ceph.com/issues/53477 [rbd mirror daemon crash]

**Comment by majianpeng** [comment]
@ideepika . i found no cache-mode/ssd w/qemu_xfstests and no cache-mode/rwl w/fio in test cases. Are you omit ? Thanks!

**Comment by ideepika** [comment]
> @ideepika . i found no cache-mode/ssd w/qemu_xfstests and no cache-mode/rwl w/fio in test cases. Are you omit ? Thanks! 

Not sure I understand, the PR is ready to merge, but blocked by GitHub, it will be merged by today or tomorrow.

**Comment by trociny** [comment]
@ideepika I think @majianpeng meant that the test subset you was running [1] did not include some tests important for this PR. May be run an additional subset with persistent cache specific tests only?

[1] https://pulpito.ceph.com/ideepika-2021-12-02_12:23:25-rbd-wip-deepika-testing-2021-12-02-1334-distro-default-smithi/

**Comment by ideepika** [comment]
aah I see, thanks @trociny @majianpeng I think due to random test picking, those tests might got missed, thanks for pointing them out!

**Comment by ideepika** [comment]
tests are fine: https://pulpito.ceph.com/ideepika-2021-12-08_04:48:12-rbd:persistent-writeback-cache-wip-deepika-testing-distro-default-smithi

**Comment by majianpeng** [comment]
> tests are fine: https://pulpito.ceph.com/ideepika-2021-12-08_04:48:12-rbd:persistent-writeback-cache-wip-deepika-testing-distro-default-smithi

hh, i'm glad to see this news.

**Comment by ideepika** [comment]
> > tests are fine: https://pulpito.ceph.com/ideepika-2021-12-08_04:48:12-rbd:persistent-writeback-cache-wip-deepika-testing-distro-default-smithi
> 
> hh, i'm glad to see this news.

I suspect the pacific backport for TestFUA be causing the failure, will test with the patch

**Pull request merged**